File: 1. 단계별로_풀어보기/36. 트리에서의 동적 계획법/3. 2533.py
import sys

sys.setrecursionlimit(1000000)

input = sys.stdin.readline
output = sys.stdout.write

#######################
#      입력 영역       #
#######################

N = int(input())
connArr = list([] for i in range(N + 1))

for _ in range(N - 1):
  U, V = map(int, input().split())
  connArr[U].append(V)
  connArr[V].append(U)

###############################
#      기타 자료구조 영역       #
###############################

ROOT_NODE = 1
parentArr = list(-1 for i in range(N + 1))
parentArr[ROOT_NODE] = 0

#############################
#      사용할 함수 영역       #
#############################

def DFS(now):
  nowO = 1
  nowX = 0

  for next in connArr[now]:
    if parentArr[next] == -1:
      parentArr[next] = now
      O, X = DFS(next)
      nowO += min(O, X)
      nowX += O

  return [nowO, nowX]

#######################
#      출력 영역       #
#######################

O, X = DFS(ROOT_NODE)

output(f"{min(O, X)}")

File: 1. 단계별로_풀어보기/36. 트리에서의 동적 계획법/3. 2533.js
// 입력
const fs = require('fs')
const inputArr = fs
  .readFileSync(0)
  .toString()
  .trim()
  .split('\n')
  .map(row => row.trim())

// 입력 파싱
const N = +inputArr[0]
const UVArr = inputArr.slice(1).map(row => row.split(' ').map(Number))

const connInfoArr = Array.from({length: N + 1}, () => [])
const parentArr = Array.from({length: N + 1}).fill(-1)

UVArr.forEach(UV => {
  const [U, V] = UV
  connInfoArr[U].push(V)
  connInfoArr[V].push(U)
})

const ROOT_NODE = 1
parentArr[ROOT_NODE] = 0

const [O, X] = DFS(ROOT_NODE)

process.stdout.write(`${Math.min(O, X)}`)

function DFS(now) {
  let nowO = 1
  let nowX = 0

  for (let next of connInfoArr[now]) {
    if (parentArr[next] === -1) {
      parentArr[next] = now
      const [O, X] = DFS(next)
      nowX += O
      nowO += Math.min(O, X)
    }
  }

  return [nowO, nowX]
}
